### PR TITLE
Add cmake to apps/gradient_autoscheduler

### PR DESCRIPTION
**Pull request by seanprime7**
`apps/gradient_autoscheduler` currently has only Makefile.

This change allows the project-wide cmake and ctest to build apps/gradient_autoscheduler and run tests.

### CODE REVIEW

**Comment by BachiLi** [comment]
LGTM but Alex should take a better look.

**Comment by vinodgro** [comment]
I was able to build the gradient_autoscheduler on Windows with this change.

**Comment by abadams** [comment]
Steven is the other reasonable reviewer for cmake stuff. I'll add him in case Alex is busy.

**Comment by steven-johnson** [comment]
FYI, building this PR on OSX and running `ctest -R run_gradient_autoscheduler_test_cpp --output-on-failure` fails with `Error: Failed to load: libgradient_autoscheduler.so: dlopen(libgradient_autoscheduler.so, 1): image not found`

File: apps/gradient_autoscheduler/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,30 @@
+add_library(gradient_autoscheduler
+            SHARED
+            GradientAutoscheduler.cpp
+            ASLog.cpp)
+target_link_libraries(gradient_autoscheduler PRIVATE "${HALIDE_COMPILER_LIB}")
+target_include_directories(gradient_autoscheduler PRIVATE "${HALIDE_INCLUDE_DIR}" "${HALIDE_TOOLS_DIR}")
+install(TARGETS gradient_autoscheduler LIBRARY DESTINATION bin)
```
Comment (steven-johnson) [inline apps/gradient_autoscheduler/CMakeLists.txt]: You need to add these lines before `install(...)`:

```
if(NOT MSVC)
  set_target_properties(gradient_autoscheduler PROPERTIES LINK_FLAGS "-rdynamic")
elseif(APPLE)
  set_target_properties(gradient_autoscheduler PROPERTIES LINK_FLAGS "-undefined dynamic_lookup")
endif()

if(NOT HALIDE_ENABLE_RTTI)
  if(NOT MSVC)
    target_compile_options(gradient_autoscheduler PRIVATE "-fno-rtti")
  else()
    target_compile_options(gradient_autoscheduler PRIVATE "/GR-")
  endif()
endif()
```

Comment (seanprime7) [reply apps/gradient_autoscheduler/CMakeLists.txt]: Okay.  I will add this to the change.

**Comment by seanprime7** [comment]
> 
> 
> FYI, building this PR on OSX and running `ctest -R run_gradient_autoscheduler_test_cpp --output-on-failure` fails with `Error: Failed to load: libgradient_autoscheduler.so: dlopen(libgradient_autoscheduler.so, 1): image not found`

I have tested it on Linux and Windows, but not on macos.  Let me try to reproduce it on macos and fix it.

**Comment by steven-johnson** [comment]
Please see https://github.com/halide/Halide/pull/4716 for my version with the changes needed for OSX -- feel free to add them here (or alternately we can move to that one)

**Comment by seanprime7** [comment]
@steven-johnson I have cherry-picked your commit into this branch.

I have one question though about the following snippet:

```
if(NOT MSVC)
  set_target_properties(gradient_autoscheduler PROPERTIES LINK_FLAGS "-rdynamic")
elseif(APPLE)
  set_target_properties(gradient_autoscheduler PROPERTIES LINK_FLAGS "-undefined dynamic_lookup")
endif()
```

Shouldn't we change this to the following?  On `APPLE`, `if(NOT MSVC)` will be `true` and `-rdynamic` would be added and the control will not go into `elseif(APPLE)` due to `else`, will it?

```
if(APPLE)
  set_target_properties(gradient_autoscheduler PROPERTIES LINK_FLAGS "-undefined dynamic_lookup")
elseif(NOT MSVC)
  set_target_properties(gradient_autoscheduler PROPERTIES LINK_FLAGS "-rdynamic")
endif()
```

File: apps/gradient_autoscheduler/CMakeLists.txt (hunk 1 above)
Comment (seanprime7) [reply apps/gradient_autoscheduler/CMakeLists.txt]: Cherry-picked your e28ccd87e57c174a1c9afd9413e5d816400f84a3 with said change.

**Comment by steven-johnson** [comment]
> I have one question though about the following snippet:

Heh, yeah, I suspect you are right -- this is code that I took from apps/autoscheduler, so maybe it's wrong there too. 

That said, probably better to leave them the same since both will be obviated soon by the cmake-modern branch.

**Comment by steven-johnson** [comment]
(still waiting on the clone-for-testing to complete)

**Comment by alexreinking** [comment]
This looks good to me for now. I'll have to update it to work on `cmake-modern`, but this still saves me some work. Thanks!

File: apps/autoscheduler/CMakeLists.txt [hunk 2]
```diff
@@ -41,8 +41,10 @@ target_link_libraries(retrain_cost_model_process
 # =======================================================
 
 # libauto_schedule
+# Note: must use MODULE here (not SHARED) to get .so (instead of .dylib) on OSX.
+# This means that this can only be opened dynamically (not linked directly), but that's ok.
 add_library(auto_schedule
-            SHARED
+            MODULE
```
Comment (alexreinking) [inline apps/autoscheduler/CMakeLists.txt]: Good catch.

**Pull request merged**